sdk/spring/spring-messaging-azure-eventhubs/src/main/java/com/azure/spring/eventhubs/core/EventHubsTemplate.java

```diff
@@ -54,19 +57,39 @@ public EventHubsTemplate(EventHubsProducerFactory producerFactory) {
         EventHubProducerAsyncClient producer = producerFactory.createProducer(destination);
         CreateBatchOptions options = buildCreateBatchOptions(partitionSupplier);
 
-        return producer.createBatch(options).flatMap(batch -> {
-            for (EventData event : events) {
-                try {
-                    batch.tryAdd(event);
-                } catch (AmqpException e) {
-                    LOGGER.error("Event is larger than maximum allowed size. Exception: " + e);
-                }
-            }
-            return producer.send(batch);
-        });
+        AtomicReference<EventDataBatch> currentBatch = new AtomicReference<>(
+                producer.createBatch(options).block());
+
+        Flux.fromIterable(events).flatMap(event -> {
+                    final EventDataBatch batch = currentBatch.get();
+                    if (batch.tryAdd(event)) {
+                        return Mono.empty();
+                    }
+
+                    return Mono.when(
+                            producer.send(batch),
+                            producer.createBatch(options).map(newBatch -> {
+                                currentBatch.set(newBatch);
+                                // Add the event that did not fit in the previous batch.
+                                if (!newBatch.tryAdd(event)) {
+                                    throw Exceptions.propagate(new IllegalArgumentException(
+                                            "Event was too large to fit in an empty batch. Max size: " + newBatch.getMaxSizeInBytes()));
+                                }
+
+                                return newBatch;
+                            }));
+                }).then()
+                .doFinally(signal -> {
+                    final EventDataBatch batch = currentBatch.getAndSet(null);
+                    if (batch != null && batch.getCount() > 0) {
+                        producer.send(batch).block();
+                    }
+                }).subscribe();
+
+        return Mono.empty();
     }
 
-    private CreateBatchOptions buildCreateBatchOptions(PartitionSupplier partitionSupplier) {
```

Comment:
Are we changing this for the test?

Reply:
yes, this can make it testable for private method is hard to mock.

Reply:
How about refactoring the code to make it testable?

Reply:
find it accept null value in test is ok, so roll it back private again. no more concern.

Reply:
refactor the mock process, then no need concern here .